### PR TITLE
SearchKit - Format footer tally values according to data type

### PR DESCRIPTION
**Pull request by colemanw**
Overview
----------------------------------------
When a SearchDisplay footer contains an average/first/last/min/max/sum of a monetary column, format that aggregate value as money.

Before
----------------------------------------
Aggregate value was presented as an unformatted integer.

After
----------------------------------------
Aggregate value is formatted.

Comments
------------
This is a reviewer's version of #30738

### CODE REVIEW

**Comment by colemanw** [comment]
@highfalutin can you review this pls?

**Comment by highfalutin** [comment]
Yes! Will do after I get past a deadline.

**Comment by highfalutin** [comment]
Ok, finally having a look at this. Code looks clean. As for how well it works -- it's an improvement.

### Most common use case

For sums of monetary amounts where all the items are in the default currency, the result is correct:

Screenshot 2024-09-12 at 10 36 30

### Less common use cases

When the monetary items have different currencies, the sum is shown in the default currency:

Screenshot 2024-09-12 at 10 23 16

I'd argue that site admins shouldn't use the "sum" tally here anyway, because it won't ever make any sense. So this is not a merge blocker.

Note that the tallies are for all items, even though the pager only shows a subset. This could be confusing, but it's not the fault of this PR.

Note also, in the screenshot above, I'm using the "count" tally on the financial type column. It works fine, however, it's a bare number, not labeled as "count". This is potentially confusing, but it's not the fault of this PR.

"List" tallies cause the footer to break. That _is_ the fault of this PR.

tally

**Comment by colemanw** [comment]
@highfalutin whew that was a doozey. Clearly it was in need of a unit test, which I've now added.

File: ext/search_kit/Civi/Api4/Action/SearchDisplay/Run.php [hunk 1]
```diff
@@ -71,10 +72,23 @@ protected function processResult(\Civi\Api4\Result\SearchDisplayRunResult $resul
         $sql = $query->getSql();
         $select = [];
         foreach ($settings['columns'] as $col) {
-          if (!empty($col['tally']['fn']) && !empty($col['key'])) {
-            $fn = \CRM_Core_DAO::escapeString($col['tally']['fn']);
-            $key = \CRM_Core_DAO::escapeString($col['key']);
-            $select[] = $fn . '(`' . $key . '`) `' . $key . '`';
+          if (!empty($col['tally']['fn']) && \CRM_Utils_Rule::mysqlColumnNameOrAlias($col['key'] ?? '')) {
```
Comment (highfalutin) [inline ext/search_kit/Civi/Api4/Action/SearchDisplay/Run.php:75]:
```php
&& \CRM_Utils_Rule::mysqlColumnNameOrAlias($col['key'] ?? '')
```

This prevents things like `financial_type_id:label` from getting tallied.

**Review by highfalutin — COMMENTED** [review]
Thanks for all the work, @colemanw ! I just tested the latest version manually, and it still works for the most plausible use case: SUM of numbers or money.

### Uncommon/unlikely use cases

#### Lists of monetary values

Now all are formatted with the _default_ currency symbol:

Screenshot 2024-09-13 at 10 12 26

Pretty hard to see how you'd get around this problem without major changes to how the tally is calculated. But also, pretty hard to see why anyone would want a list of all values in the footer. (Unless, maybe, you could just list `DISTINCT` values.)

#### Tallies of label fields

For example, `financial_type_id:label`. The latest version of this PR makes tallies of that field blank. However, again, it's hard to see why anyone would want a List/First/Count of Financial Types in the footer, unless maybe you could count/list distinct financial types.

**Comment by highfalutin** [comment]
My vote is to merge this PR as is and move on.

**Pull request merged**